### PR TITLE
Display user facing error when direct upload fails

### PR DESCRIPTION
**Pull request by JPrevost**
## Status
**READY**

#### What does this PR do?

If a direct upload fails, display a user facing message. Our js exception will alert us of the problems, this is just a note to users that we know it happened.

#### How can a reviewer manually see the effects of these changes?

PR build will be configured with a broken S3 connection.

screen shot 2018-03-28 at 11 00 19 am

#### What are the relevant tickets?
https://mitlibraries.atlassian.net/browse/TI-71

#### Requires Database Migrations?
NO

#### Includes new or updated dependencies?
NO

### CODE REVIEW

File: app.json [hunk 1]
```diff
@@ -21,6 +21,9 @@
     "HEROKU_APP_NAME": {
       "required": true
     },
+    "JS_EXCEPTION_LOGGER_KEY": {
```
Comment (thatandromeda) [inline app.json]: I take it rollbar is smart enough to notice when things are coming from the PR builds and keep those separate from prod logs?

File: app/assets/javascripts/direct_uploads.js [hunk 2]
```diff
@@ -29,8 +29,14 @@ addEventListener("direct-upload:error", event => {
   event.preventDefault()
   const { id, error } = event.detail
   const element = document.getElementById(`direct-upload-${id}`)
+  Rollbar.critical("ActiveStorage Direct Upload Failed.", event);
   element.classList.add("direct-upload--error")
   element.setAttribute("title", error)
+  element.insertAdjacentHTML("afterend", `
```
Comment (thatandromeda) [inline app/assets/javascripts/direct_uploads.js]: @frrrances is this going to be difficult for you to maintain if you want to change the UX? Do we need to document this somewhere?

File: app/views/layouts/_js_exception_handler.erb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
...captureUncaughtExceptions:o,captureUnhandledRejections:t,wrapGlobals:a}},function(r,e){"use strict";function o(r,e){this.impl=r(e,this),this.options=e,n(o.prototype)}function n(r){for(var e=function(r){return function(){var e=Array.prototype.slice.call(arguments,0);if(this.impl[r])return this.impl[r].apply(this.impl,e)}},o="log,debug,info,warn,warning,error,critical,global,configure,handleUncaughtException,handleUnhandledRejection,_createItem,wrap,loadFull,shimId,captureEvent,captureDomContentLoaded,captureLoad".split(","),n=0;n<o.length;n++)r[o[n]]=e(o[n])}o.prototype._swapAndProcessMessages=function(r,e){this.impl=r(this.options);for(var o,n,t;o=e.shift();)n=o.method,t=o.args,this[n]&&"function"==typeof this[n]&&("captureDomContentLoaded"===n||"captureLoad"===n?this[n].apply(this,[t[0],o.ts]):this[n].apply(this,t));return this},r.exports=o},function(r,e){"use strict";r.exports=function(r){return function(e){if(!e&&!window._rollbarInitialized){r=r||{};for(var o,n,t=r.globalAlias||"Rollbar",a=window.rollbar,l=function(r){return new a(r)},i=0;o=window._rollbarShims[i++];)n||(n=o.handler),o.handler._swapAndProcessMessages(l,o.messages);window[t]=n,window._rollbarInitialized=!0}}}}]);
```
Comment (thatandromeda) [inline app/views/layouts/_js_exception_handler.erb]: I wonder what this actually sends/surveils.

Comment (JPrevost) [reply app/views/layouts/_js_exception_handler.erb]: You can see some example logs from the test reporting. It's definitely comprehensive and once we have policy around this we can review it in that context.

File: app.json (hunk 1 above)
Comment (JPrevost) [reply app.json]: The request url in the of the problem will be the PR build. If it becomes difficult to distinguish with just that we can adjust.

**Comment by JPrevost** [comment]
@frrrances PR build is intentionally borked. Upload then submit an otherwise valid form to see the error kick in.

File: app/assets/javascripts/direct_uploads.js [hunk 4]
```diff
@@ -29,8 +29,14 @@ addEventListener("direct-upload:error", event => {
...
+    <div id="direct-upload-${id}" class="alert alert-banner error">
+      An error has occured. Sorry for the incovenience. System engineers have been alerted. Please try again later. If problems persist, please contact etheses-admin@mit.edu for assistance.
```
[the remark this answers is not in the document]
Comment (JPrevost) [reply app/assets/javascripts/direct_uploads.js]: At this time, it is indeed just expected to happen when we can't communicate properly with S3. I'll update the PR to use your suggested words. Thx.

**Comment by JPrevost** [comment]
a11y concerns will be address after accessibility review is complete.

**Pull request merged**